src/middleware/redux.ts

```diff
@@ -1,62 +1,59 @@
-import { GetState, SetState, State, StoreApi } from '../vanilla'
+import { State, StateCreator, StoreMutatorIdentifier } from '../vanilla'
 import { NamedSet } from './devtools'
 
-type DevtoolsType = {
-  prefix: string
-  subscribe: (dispatch: any) => () => void
-  unsubscribe: () => void
-  send: (action: string, state: any) => void
-  init: (state: any) => void
-  error: (payload: any) => void
-}
+type Write<T extends object, U extends object> = Omit<T, keyof U> & U
+type Cast<T, U> = T extends U ? T : U
 
-/**
- * @deprecated Use `Mutate<StoreApi<T & { dispatch: (a: A) => A }>, [["zustand/redux", A]]>`.
- * See tests/middlewaresTypes.test.tsx for usage with multiple middlewares.
- */
-export type StoreApiWithRedux<
+type WithRedux<S, A> = Write<Cast<S, object>, StoreRedux<Cast<A, Action>>>
+
+type Redux = <
   T extends State,
-  A extends { type: unknown }
-> = StoreApi<T & { dispatch: (a: A) => A }> & {
-  dispatch: (a: A) => A
-  dispatchFromDevtools: boolean
-}
+  A extends Action,
+  Cms extends [StoreMutatorIdentifier, unknown][] = []
+>(
+  reducer: (state: T, action: A) => T,
+  initialState: T
+) => StateCreator<Write<T, ReduxState<A>>, Cms, [['zustand/redux', A]]>
 
 declare module '../vanilla' {
   interface StoreMutators<S, A> {
     'zustand/redux': WithRedux<S, A>
   }
 }
 
-interface StoreRedux<A extends Action> {
-  dispatch: (a: A) => A
-  dispatchFromDevtools: true
-}
-
 interface Action {
   type: unknown
 }
 
-type Write<T extends object, U extends object> = Omit<T, keyof U> & U
-type Cast<T, U> = T extends U ? T : U
+interface ReduxState<A extends Action> {
+  dispatch: StoreRedux<A>['dispatch']
+}
 
```

Comment:
Move up please. (sorry to bother..)

Reply:
Yeah don't worry, feel free to tweak things yourself too if you wish.